apps/omg_watcher/lib/omg_watcher/block_validator.ex

```diff
@@ -64,4 +73,71 @@ defmodule OMG.Watcher.BlockValidator do
       end
     end)
   end
+
+  @spec number_of_transactions_within_limit([Transaction.Signed.tx_bytes()]) :: :ok | {:error, atom()}
+  defp number_of_transactions_within_limit(transactions) when length(transactions) == 0, do: {:error, :empty_block}
+
+  defp number_of_transactions_within_limit(transactions) when length(transactions) > @transaction_upper_limit do
+    {:error, :transactions_exceed_block_limit}
+  end
+
+  defp number_of_transactions_within_limit(_transactions), do: :ok
+
+  @spec verify_no_duplicate_inputs([Transaction.Recovered.t()], map()) :: {:ok, map()}
+  defp verify_no_duplicate_inputs([], input_set), do: {:ok, input_set}
```

Comment:
can this be simplified by something like 
```
defp verify_no_duplicate_inputs(transactions) do
          all_inputs = Enum.flat_map(transactions, &Transaction.get_inputs/1)
          uniq_inputs = Enum.uniq_by(all_inputs, &Position.encode/1)
          
          case length(all_inputs) == length(uniq_inputs) do
            true -> :ok
            false -> {:error, :block_duplicate_inputs}
          end    
        end
```
Not sure if it's faster

Reply:
Comparing both versions: 

t = number of transactions
i = number of inputs 

** Proposed version **

map to inputs: O(t)
flatten: O(t)
encode position: O(i)
`uniq_by`: O(i) [best case]

perf: O(2t + 2i) 

** Current version ** 

get_inputs: O(t)
endode: O(i)
check if in accumulator map: O(i)

perf: O(t + 2i)

pasting my version for ease
```  
defp verify_no_duplicate_inputs([], input_set), do: {:ok, input_set}

  defp verify_no_duplicate_inputs([transaction | rest], input_set) do
    current_input_positions = transaction |> Transaction.get_inputs() |> Enum.map(&Position.encode/1)

    current_input_positions
    |> Enum.any?(fn input_position -> Map.has_key?(input_set, input_position) end)
    |> case do
      true ->
        {:error, :block_duplicate_inputs}

      false ->
        new_input_set = current_input_positions |> Map.new(fn pos -> {pos, true} end) |> Map.merge(input_set)
        verify_no_duplicate_inputs(rest, new_input_set)
    end
```

@InoMurko would appreciate your insights on this one! @mederic-p 's definitely looks better though.

Reply:
pick the one that's easier to read and understand. we can make it fast if we need to.